ethcore/sync/src/chain/mod.rs

```diff
@@ -769,10 +827,37 @@ impl ChainSync {
 
     /// Updates transactions were received by a peer
     pub fn transactions_received(&mut self, txs: &[UnverifiedTransaction], peer_id: PeerId) {
-        if let Some(peer_info) = self.peers.get_mut(&peer_id) {
-            peer_info
-                .last_sent_transactions
-                .extend(txs.iter().map(|tx| tx.hash()));
+        // Remove imported txs from all request queues
+        let imported = txs.iter().map(|tx| tx.hash()).collect::<H256FastSet>();
+        for (pid, peer_info) in &mut self.peers {
+            peer_info.unfetched_pooled_transactions = peer_info
+                .unfetched_pooled_transactions
+                .difference(&imported)
+                .copied()
+                .collect();
+            if *pid == peer_id {
+                match GetPooledTransactionsReport::generate(
+                    std::mem::replace(&mut peer_info.asking_pooled_transactions, Vec::new()),
```

Comment:
You might rather `peer_info.asking_pooled_transactions.drain()` here and change the signature of `generate` to take an `Iterator` as well to avoid re-allocations.